File: gauss.py
"""
$ python gauss.py
"""
import numpy as np
import time

class Gauss:
  #n = np.int64(1000)
  #a = np.zeros((n, n), dtype=np.float64)
  #b = np.zeros(n, dtype=np.float64)
  def __init__(self):
    Gauss.n = 1000
    self.a = np.zeros((Gauss.n, Gauss.n))
    self.b = np.zeros(Gauss.n)
  def set(self, a, b):
    for i in range(Gauss.n):
      for j in range(i, Gauss.n):
        self.a[i][j] = a
      self.b[i] = b
    #self.a = [[1, 1, 1, 1, 1], [0, 1, 1, 1, 1], [0, 0, 1, 1, 1], [0, 0, 0, 1, 1], [0, 0, 0, 0, 1]]
    #self.b = [1, 1, 1, 1, 1]
  def solve(self):
    for k in range(Gauss.n - 1):
      for i in range(k + 1, Gauss.n):
        s = self.a[i][k] / self.a[k][k]
        for j in range(k + 1, Gauss.n):
          self.a[i][j] -= s * self.a[k][j]
        self.b[i] -= s * self.b[k]
    for i in range(Gauss.n - 1, - 1, - 1):
      for j in range(i + 1, Gauss.n):
        self.b[i] -= self.a[i][j] * self.b[j]
      self.b[i] /= self.a[i][i]
  def print(self):
    for i in range(len(self.b)):
      print(f"x[{i}] = {self.b[i]:.10E}")

gauss = Gauss()
gauss.set(1, 1)
start = time.time()
gauss.solve()
stop = time.time()
gauss.print()
print(f"{(stop - start):.6f}sec")

File: gauss.jl
#=
$ julia gauss.jl
=#

using Printf

struct Gauss
  a::Vector{Vector{Float64}}
  b::Vector{Float64}
  #Gauss(n) = new([zeros(Float64, n) for i = 1 : n], zeros(Float64, n))
#=
  function Gauss(n::Int)
    a = [zeros(Float64, n) for i = 1 : n]
    b = zeros(Float64, n)
    new(a, b)
  end
=#
end

function Gauss(n::Int)
  a = [zeros(Float64, n) for i = 1 : n]
  b = zeros(Float64, n)
  Gauss(a, b)
end

function set!(gauss::Gauss, a::Float64, b::Float64)
  n = length(gauss.a)
  for i = 1 : n
    for j = i : n
      gauss.a[i][j] = a
    end
    gauss.b[i] = b
  end
end

function solve!(gauss::Gauss)
  n = length(gauss.a)
  for k = 1 : n - 1
    for i = k + 1 : n
      s = gauss.a[i][k] / gauss.a[k][k]
      for j = k + 1 : n
        gauss.a[i][j] -= s * gauss.a[k][j]
      end
      gauss.b[i] -= s * gauss.b[k]
    end
  end
  for i = n : - 1 : 1
    for j = i + 1 : n
      gauss.b[i] -= gauss.a[i][j] * gauss.b[j]
    end
    gauss.b[i] /= gauss.a[i][i]
  end
end

function print(gauss::Gauss)
  for i = 1 : length(gauss.b)
    @printf("x[%d] = %.10E\n", i, gauss.b[i])
  end
end

const gauss = Gauss(1000)
set!(gauss, 1.0, 1.0)
@time begin
  solve!(gauss)
end
#print(gauss)